orbstation/antagonists/traitor/objectives/final_objective/malware_injection.dm

```diff
@@ -0,0 +1,432 @@
+		user.visible_message(span_danger("[user] hunches over [src] and begins typing away!"), span_notice("You start typing on [src] to resume the malware injection."))
+		resuming = TRUE
+		if(!do_after(user, INJECTOR_HACK_TIME, src, IGNORE_SLOWDOWNS, extra_checks = CALLBACK(src, .proc/inject_check)))
+			resuming = FALSE
+			return
+		resuming = FALSE
+		to_chat(user, span_boldnotice("Resuming malware injection."))
+		// time to go loud!
+		trigger_alarm(user) // set off the alarm, close and lock all doors in the area, alert the AI in big bold letters
+		announce_corruption()
+		advance_corruption_stage(user)
+		return
+	current = select_active_ai(user) // z-level does not matter for this
+	if(!current)
+		to_chat(user, span_alert("No active AIs detected!"))
+	else
+		to_chat(user, span_notice("[current.name] selected for malware injection."))
+
+/obj/machinery/computer/malware/attackby(obj/item/O, mob/living/user, params)
+	if(istype(O, /obj/item/malware_injector))
+		var/obj/item/malware_injector/injector = O
+		if(machine_stat & (NOPOWER|BROKEN|MAINT))
+			return
+		if(!istype(user))
+			return
+		if(!user.mind.has_antag_datum(/datum/antagonist/traitor))
+			to_chat(user, span_warning("You have no idea how to use [injector]."))
+			return
+		if(injector.spent)
+			to_chat(user, span_alert("[injector] is no longer functional."))
+			return
+		if(injector.hacking)
+			to_chat(user, span_alert("[injector] is already hacking something else."))
+			return
+		if(length(injector.hacked_apcs) < APCS_REQUIRED_FOR_MALF)
+			to_chat(user, span_alert("[injector] has not hacked enough APCs to inject the malware. [APCS_REQUIRED_FOR_MALF - length(injector.hacked_apcs)] more needed."))
+			return
+		if(!current)
+			to_chat(user, span_alert("You haven't selected anything to inject malware into!"))
+			return
+		if(!inject_check())
+			to_chat(user, span_alert("Upload failed! Check to make sure [current.name] is functioning properly."))
+			current = null
+			return
+		if(!objective_owner || !req_area) // we need to tell the computer who owns the malware objective and what the required area is
+			get_objective_info(user)
+		var/area/comp_area = get_area(src)
+		if(user.mind != objective_owner) // you don't have the malware injection objective
+			to_chat(user, span_alert("Access denied."))
+			return
+		if(!istype(comp_area, req_area))
+			to_chat(user, span_alert("Upload failed! Unable to establish connection to [current.name] in this area."))
+			return
+		begin_injection(injector, user)
+	else
+		return ..()
+
+/obj/machinery/computer/malware/screwdriver_act(mob/living/user, obj/item/I)
+	if(injecting)
+		to_chat(user, span_alert("ERROR: Malware injection in progress."))
+		return FALSE
+	return ..()
+
+///Gets information about the malware injection objective's owner and the required area the console needs to be in. This should only run once.
+/obj/machinery/computer/malware/proc/get_objective_info(mob/living/user)
```

Comment:
I can understand the need to keep this, but we are still calling it three times which does not seem very efficient.
Can't we just call this once when setting up the computer and store the values we want then?

Reply:
In practice it should only call it once, which is why it has an if(!objective_owner || !req_area) check before each one. Once get_objective_info() has been run once by the traitor with the objective, it will have that information so it won't call the proc anymore.